Heroku collaborator module

marns93 opened the pull request:
##### SUMMARY

This module allows managing Heroku app collaborators. Heroku doesn't offer an interface that provides overview across all organization apps, therefore managing collaborators without Ansible is difficult and error-prone. The module solves this problem.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
heroku_collaborator

marns93 committed:
New module: heroku_collaborator

marns93 commented:
ready_for_review

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/python
+
+# Copyright: Ansible Project
```
```
Copyright: (c) 2018, Ansible Project
```

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+    required: true
+  state:
+    description:
+      - Create or remove the heroku collaborator
```
Add description like with all possible combinations -
```
If set to C(present) and heroku user is already collaborator, then ...
```

marns93 replied:
Thank you, we added requested documentation.

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+    default: "present"
+notes:
+  - C(HEROKU_API_KEY) and C(TF_VAR_HEROKU_API_KEY) env variable can be used instead setting c(api_key).
+  - If you use I(--check), you can also pass the I(-v) flag to see affected apps in msg.
```
New person will be confused when you refer to `msg`, adding a brief description will help.

marns93 replied:
Here, we tried to add more documentation, but after several tries couldn't get YAML to validate. If you think that our current suggestion is still not acceptable, we would appreciate your help.

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+    - { user: 'x.y@example.com', apps: ["heroku-example-app"] }
+'''
+
+RETURN = ''' # '''
```
Return is not compulsory but will help end user to understand return value of module.

marns93 replied:
The documentation says:

```
If your module doesn't return anything (apart from the standard returns), you can use ``RETURN = ''' # '''``.
```

Our module indeed doesn't return anything apart from the standard returns (`msg`). If you think that we should document something we would appreciate further hints.

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+
+import os
+
+DEPENDENCY_CHECK = True
```
Might want to change to `HAS_HEROKU` to match other Ansible modules code.

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            api_key=dict(default=os.getenv('HEROKU_API_KEY', os.getenv('TF_VAR_HEROKU_API_KEY')), type='str'),
```
Use `env_fallback`

```
from ansible.module_utils.basic import env_fallback
...
api_key=dict(fallback=(env_fallback, ['HEROKU_API_KEY', TF_VAR_HEROKU_API_KEY']), type='str', no_log=True),
```

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+        module.fail_json(msg='Heroku authentication failure, please check your API Key')
+
+    # Prevent API key from appearing in debug logs
+    del module.params['api_key']
```
`no_log=True` is argument spec will handle this.

Akasurde commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,142 @@
+        supports_check_mode=True
+    )
+
+    if not DEPENDENCY_CHECK:
```
One suggestion, it will be helpful to create a common library under module_utils so that in future all Heroku module will use same like [DigitalOcean](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/digital_ocean.py) or [influxdb](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/influxdb.py).

Akasurde reviewed (changes requested):
Thanks for PR.

marns93 commented:
@Akasurde thank you very much for your review and hints, we have implemented everything to the best of our knowledge.

ready_for_review

Akasurde commented:
LGTM. We need one more person to take a look and then we are good to go.

zyv commented:
Hi @Akasurde, do you think you could help us to find a second reviewer? Quite honestly we're not quite sure as to how to proceed. Many thanks!

Akasurde commented:
@resmo @ryansb @willthames Could you please provide your valuable reviews ?

resmo commented on lib/ansible/module_utils/heroku.py:
```diff
@@ -0,0 +1,38 @@
+# Copyright: (c) 2018, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
blocker: code under module_utils/* has to be BSD licensed.

resmo commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,130 @@
+  - If set to C(present) and heroku user is not collaborator, then add user to app.
+  - If set to C(absent) and heroku user is collaborator, then delete user from app.
+author:
+  - Marcel Arns <marcel.arns@moneymeets.com>
```
not a blocker, but please replace your mail with your github handle.

resmo commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,130 @@
+      - Suppress email invitation when creating collaborator
+    type: bool
+    required: false
+    default: "False"
```
not a blocker, but it looks better in the docs if you use (yes/no) for type bool.

resmo commented on lib/ansible/modules/cloud/heroku/heroku_collaborator.py:
```diff
@@ -0,0 +1,130 @@
+  api_key:
+    description:
+      - Heroku API key
+    required: false
```
not a blocker, but requried=false is the default and can be omitted.

resmo reviewed (changes requested):
Wrong license in module_utils is a blocker. Few minor things noted but would not hold it back. looks good to me when module_utils code is bsd licensed.

marns93 commented:
@resmo thank you for the review, we have addressed your comments in 6f01c92. We have also rebased the PR against latest devel and squashed fixup commits. We have noticed that changelogs are apparently automatically generated now, so we have removed the changelog commit. If we've got it wrong, we will appreciate further guidance on your part. Many thanks!

resmo commented:
shipit
